src/insi-client.service.ts

```diff
@@ -116,31 +116,54 @@ export class INSiClient {
       }
       this._soapClient.clearSoapHeaders();
     }
-    return {
-      ...this._getFetchResponseFromRawSoapResponse(rawSoapResponse, requestId),
-      failedRequests: failedRequests,
-    };
+    return responses;
+  }
+
+  private _callFetchFromIdentityTraits(requestId: string, soapBody: INSiPersonSoapBody): Promise<INSiFetchInsResponse> {
+    const { method } = INSiSoapActions[INSiSoapActionsName.FETCH_FROM_IDENTITY_TRAITS];
+    return new Promise<INSiFetchInsResponse>(async (resolve, reject) => {
+      this._soapClient[`${method}`](soapBody, (err: any, result: any, rawResponse: string, soapHeader: any, rawRequest: string) => {
+        if (err?.response?.status === 500 && err.body) {
+          this._getServiceErrorFromXML(rawResponse);
+          resolve({
+            status: INSiServiceRequestStatus.FAIL,
+            requestId,
+            response: null,
+            responseBodyAsJson: null,
+            responseBodyAsXml: rawResponse,
+            requestBodyAsXML: rawRequest,
+            error: this._getServiceErrorFromXML(rawResponse),
+          });
+        } else if (err) {
+          reject(err);
+        } else {
+          resolve({
+            status: INSiServiceRequestStatus.SUCCESS,
+            requestId,
+            response: InsiHelper.formatFetchINSResult(result),
+            responseBodyAsJson: InsiHelper.changeInsHistoToArray(result),
+            responseBodyAsXml: rawResponse,
+            requestBodyAsXML: rawRequest,
+            error: null,
+          });
+        }
+      });
+    });
   }
 
   private _setSoapHeaders(requestId: string): void {
     const { header } = INSiSoapActions[INSiSoapActionsName.FETCH_FROM_IDENTITY_TRAITS];
     this._setDefaultHeaders();
     this._soapClient.addSoapHeader(header, 'Action', 'wsa', 'http://www.w3.org/2005/08/addressing');
     this._soapClient.addSoapHeader({ MessageID: `uuid:${requestId}` }, 'MessageID', 'wsa', 'http://www.w3.org/2005/08/addressing');
-    
   }
 
-  private _getFetchResponseFromRawSoapResponse(rawSoapResponse: any, requestId: string): INSiFetchInsResponse {
-    const [rawResponse, responseAsXMl, , requestAsXML] = rawSoapResponse;
-    if (rawResponse?.INDIVIDU?.INSHISTO && !_.isArray(rawResponse?.INDIVIDU?.INSHISTO)) {
-      rawResponse.INDIVIDU.INSHISTO = [rawResponse.INDIVIDU.INSHISTO];
-    }
+  private _getServiceErrorFromXML(xml: any): any {
     return {
-      requestId,
-      rawBody: rawResponse,
-      body: InsiHelper.formatFetchINSRawResponse(rawResponse),
-      bodyAsXMl: responseAsXMl,
-      requestBodyAsXML: requestAsXML,
+      siramCode: xml.match(/(<S:Subcode><S:Value>S:)(.*)(<\/S:Value>)/)[2],
```

Comment:
prefer proper XML parsing over regex. regex can break more easily.

Reply:
I will check this solution when i'm done with front and back.